storage/innobase/include/trx0trx.h

```diff
@@ -779,13 +779,62 @@ struct trx_lock_t {
 	bool		start_stmt;
 };
 
+/** Logical first modification time of a table in a transaction */
+class trx_mod_table_time_t
```

Comment:
What are other use cases of this class apart from detecting historical data modifications? The logic is too complex just for use case of updating TRT table.

Reply:
This class encapsulates two pairs of `undo_no_t`. The first one is what was added in the bug fix
[MDEV-14477 InnoDB update_time is wrongly updated after partial rollback or internal COMMIT](https://jira.mariadb.org/browse/MDEV-14477).

The second `undo_no_t` is what actually allows us to replace the `vers_update_trt` flags. Yes, the logic is somewhat complex, but in my opinion, it still is less error-prone than the approach of updating a `vers_update_trt` flag in multiple places. With that kind of approach, some adjustment could be forgotten. And the flag will be out of sync with the actual transaction state in a few corner cases. It could be even more fragile with cross-engine transactions where multiple transactional engines support system-versioned columns.

I think that it is cleaner design that each transaction can be queried at any point of time whether they have modified any system-versioned data. With this patch, the `trx_t::mod_tables` is guaranteed to be in sync with the transaction state, as follows:
* Before each transactional change, after the undo log record was written successfully, `trx_undo_report_row_operation()` will update the `trx_t::mod_tables`.
* Each rollback (whether an internal row-rollback in InnoDB, or `ROLLBACK TO SAVEPOINT` or rollback to the start of a statement) will update the `trx_t::mod_tables`.